backend/routes/forms/unittesting.py

```diff
@@ -0,0 +1,132 @@
+import ast
+from collections import namedtuple
+from itertools import count
+from textwrap import indent
+from typing import Optional
+
+import httpx
+
+from backend.constants import SNEKBOX_URL
+from backend.models import FormResponse, Form
+
+with open("resources/unittest_template.py") as file:
+    TEST_TEMPLATE = file.read()
+
+
+UnittestResult = namedtuple("UnittestResult", "question_id return_code passed result")
+
+
+def filter_unittests(form: Form) -> Form:
+    """
+    Replace the unittest data section of code questions with the number of test cases.
+
+    This is used to redact the exact tests when sending the form back to the frontend.
+    """
+    for question in form.questions:
+        if question.type == "code" and "unittests" in question.data:
+            question.data["unittests"] = len(question.data["unittests"])
+
+    return form
+
+
+def _make_unit_code(units: dict[str, str]) -> str:
+    """Compose a dict mapping unit names to their code into an actual class body."""
+    result = ""
+
+    for unit_name, unit_code in units.items():
+        result += f"\ndef test_{unit_name.lstrip('#')}(unit):\n{indent(unit_code, '    ')}"
+
+    return indent(result, "    ")
+
+
+def _make_user_code(code: str) -> str:
+    """Compose the user code into an actual string variable."""
+    # Make sure that we we escape triple quotes in the user code
+    code = code.replace('"""', '\\"""')
```

Comment:
This is still susceptible to escape with something like `\\\"\"\"`. If this is only done to prevent it from escaping the code block on the next line, then just convert all inputted quotes to single quotes. That will still break user's code in some instances, but that's much better than allowing someone to escape quotes on our system.